### ISSUE TITLE
Blank window when selecting aerial in the screensaver prefpane

### ISSUE THREAD

**Issue by hachirotahoshino** [JohnCoates/Aerial#696]
#### General troubleshooting tips

Before logging an issue please check that:
- [x 1.4.6] You have the latest version installed (There may be a beta version that fixes your issue), see here for the latest releases and bug fixes: https://github.com/JohnCoates/Aerial/releases
- [x not exactly this and workaround of other haven't worked ] Your issue isn't already mentioned in our [issues](https://github.com/JohnCoates/Aerial/issues). You may find a workaround there or a similar request already made.
- [x haven't seen ] Your problem isn't mentioned in the [troubleshooting section](https://github.com/JohnCoates/Aerial#troubleshooting).

If none of this fixes your issue, tell us about the problem you are experiencing or the feature you'd like to request.

#### Required information
In order to help us sort your issue, we ask that you provide the following information:
- [ ] Mac model: Mac Pro 5.1 mid 2010
- [ ] macOS version: 10.14.2
- [ ] Monitor setup: 1x4k monitor

If appropriate, please enable `Debug mode` and `Log to disk` in `Advanced` tab and replicate your bug, then attach the `AerialLog.txt` file (You can access this file through the Advanced tab).

#### Description of issue / Feature request

Hi, when i select aerial i have a blank pref pane.

![capture d ecran 2018-12-27 a 09 25 12](https://user-images.githubusercontent.com/38068857/50472495-5586ad00-09b9-11e9-93aa-f856fbd347d7.jpg)

end november / early december aerial prefpane even crashed repeatedly after moving the cache folder. I removed it and forgot it but now i wanted to try again, removed all prefs, caches, try to change cache path in plist... nothing.

here is latest crash log from that time:

[ScreenSaverEngine_2018-12-06-160900_HackThalo.crash.zip](https://github.com/JohnCoates/Aerial/files/2711914/ScreenSaverEngine_2018-12-06-160900_HackThalo.crash.zip)

thanks for your help

**Comment by glouel** [comment]
@hachirotahoshino first, really sorry about that !
According to the log, it looks like you are running version 1.4.4 ? Did you try version 1.4.6 ? There were many changes since 1.4.4 and many bugs fixed. 

See here : 
https://github.com/JohnCoates/Aerial/releases/latest

I would suggest you upgrade to 1.4.6 and let me know if you still see the issue. Your log seems to point to a crash when downloading something, can you make sure you don't have a firewall or something blocking your network connections ? 

My guess is it fails downloading the JSONs for some reason. Please let me know what happens with 1.4.6 and we'll try to fix that. Thanks !

**Comment by hachirotahoshino** [comment]
Hi @glouel In 1.4.4 i had the crash, now i don't have crash but blank pref pane as in the picture. 
I don't have anything blocking outgoing connections!

Hi have checked a similar topic before and tried this link in safary without any issue:

![capture d ecran 2018-12-27 a 15 42 02](https://user-images.githubusercontent.com/38068857/50483900-0d35b200-09ee-11e9-88c2-70d9f091b8b7.jpg)

without crash, so without crash log, how can i help you debug?

**Comment by glouel** [comment]
@hachirotahoshino Thanks a lot for clarifying ! I'll make a special build with forced logs for you so we can find the issue. Will get back to you later today.

**Comment by glouel** [comment]
@hachirotahoshino sorry for the multiple message, you can find the build here : 
https://github.com/glouel/Aerial/releases/download/v1.4.7test2/Aerial.saver.zip

Please note, if you delete your .plist (/Users/YOURUSERNAME/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist), starting with 1.4.6, Aerial will do the following things at startup to look/create the cache folder: 

- Look if /Library/Caches/Aerial exists, if so use that
- Look if /Users/YOURUSERNAME/Library/Caches/Aerial exists, if so use that
- Create /Library/Caches/Aerial and use that 

This build has forced logs, so it will create AerialLog.txt and paste logs in the Cache folder computed above, if you can send me that log after running it once (feel free to delete the old one before if you had one) it would be helpful. Again sorry for your issue, we'll figure this out !

Also, you may see a prompt for updates with this version should it work, please ignore those too and don't auto update :)

**Comment by hachirotahoshino** [comment]
hi, so first i've tried with aerial in console while launching aerial:

the results i have are these:

`*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'working directory doesn't exist.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff3743cff5 __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff63594efb objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff3743ce27 +[NSException raise:format:] + 193
	3   Foundation                          0x00007fff3974ffed -[NSConcreteTask launchWithDictionary:error:] + 1808
	4   Aerial                              0x000000010533a7d1 $S6Aerial17DownloadOperationC10urlSession_4task20didCompleteWithErrorySo12NSURLSessionC_So0K4TaskCs0J0_pSgtFTf4dnnn_n + 1121
	5   Aerial                              0x000000010533ab35 $S6Aerial15DownloadManagerC10urlSession_4task20didCompleteWithErrorySo12NSURLSessionC_So0K4TaskCs0J0_pSgtFTf4dnnn_n + 309
	6   Aerial                              0x00000001053388f6 $S6Aerial15DownloadManagerC10urlSession_4task20didComplet<…>`

`[C3 Hostname#eda221fd:443 tcp, url: https://sylvan.apple.com/Aerials/2x/entries.json, tls] cancelled
	[C3.1 <private> 192.168.69.252:49157<->IPv4#65dac0ce:443]
	Connected Path: satisfied (Path is satisfied), interface: en2, ipv4, dns
	Duration: 0.100s, DNS @0.000s took 0.002s, TCP @0.005s took 0.021s, TLS took 0.053s
	bytes in/out: 13742/903, packets in/out: 14/3, rtt: 0.018s, retransmitted packets: 0, out-of-order packets: 0`

`AerialError: <private>`

`-[ScreenSaverEngine _initAndGetModule] -- Found module in defaults: Aerial`

![capture d ecran 2018-12-27 a 17 12 43](https://user-images.githubusercontent.com/38068857/50486520-a5d22f00-09fa-11e9-900f-76b05833cfe5.jpg)

**Comment by glouel** [comment]
@hachirotahoshino interesting, it looks as if the cache directory doesn't exist/wasn't created, and it probably fails to untar the tvos12 json bundle. 

Let me know what happens with the other build, particularly with your cache directory location, I think the issue is around this (it could be that the path is write protected or your user doesn't have access to it, something like that).

**Comment by hachirotahoshino** [comment]
after cleaning the caches and prefs (strangely there was no Aerial folder in neither of global nor user Library/Caches)

installed the new one

relaunch

![capture d ecran 2018-12-27 a 17 16 45](https://user-images.githubusercontent.com/38068857/50486746-aae3ae00-09fb-11e9-802c-934d520e6d08.jpg)

and same issues in console.

I went to check for logs in /L/Caches or ~/L/Caches without success. no. Aerial. Folder. at. all. :)

this may be that

**Comment by hachirotahoshino** [comment]
Yes i've deleted the plist in ~/LPreferences/byHost shouldn't i? i thought as clean install. :-/

I just tried creating Aerial folder in both /L/C and ~/L/C 

same issue

[aerial.mov.zip](https://github.com/JohnCoates/Aerial/files/2712885/aerial.mov.zip)

il start aerial, ask for the automatic update but directly goes to empty window (the message for auto update was so fast i had to record it to read it :P )

**Comment by hachirotahoshino** [comment]
ok i'll check permission

**Comment by hachirotahoshino** [comment]
`HackThalo:~ ssm$ ls -ld /Library/Caches/Aerial/
drwxrwxrwx  2 ssm  admin  64 27 déc 17:24 /Library/Caches/Aerial/
HackThalo:~ ssm$ ls -ld Library/Caches/Aerial/
drwxrwxrwx  2 ssm  staff  64 27 déc 17:23 Library/Caches/Aerial/`

**Comment by glouel** [comment]
Yes you did alright deleting the plist, I read your answer too fast, sorry about that.

I would suggest removing  /Library/Caches/Aerial and only keep  ~/Library/Caches/Aerial and try again ? Sorry about this, trying to understand what happens here.

**Comment by hachirotahoshino** [comment]
Tried with only /L/C/Aerial/ , and same issue!

Please don't be sorry, software is great and free and you're helping <3

If it help for me it happen since i deleted the folder where i wanted it to store files and placed it on another hard drive.

i thought it would just putting me an error it can't find it

then i've done so much things to uninstall reinstall that i may have tempered something :)

**Comment by glouel** [comment]
Just to be clear, I'd suggest:
removing /Library/Caches/Aerial (global one)
keep ~/Library/Caches/Aerial (in your user directory) 

And try again ?

Just to make sure, did you happen to have moved your whole user directory (not just the Aerial cache) ? 

I'm a bit puzzled about what could be happening, especially since you removed the plist. I'll think of something ;)

**Comment by hachirotahoshino** [comment]
same issue :)

nop nop didn't touch my user directory. everything's fine for the rest

on 1.4.4, when it crashed, i started by removing plist then caches/aerial folder(s) but it kept crashing, then i didn't use aerial for some time and removed it then today i d/l 1.4.6 installed, and i just had this situation

i only touched the "supposedly" aerial files, nothing else. that's also the reason i opened this bug report, because i thought i missed a setting / plist file somewhere else. :)

**Comment by glouel** [comment]
Please give this build a shot : 
https://github.com/glouel/Aerial/releases/download/v1.4.7test3/Aerial.saver.zip

It will create AerialLog.txt on your desktop instead, which should be safe permission wise, and hopefully lets me find the issue. Thanks.

**Comment by hachirotahoshino** [comment]
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2713101/AerialLog.txt)

**Comment by hachirotahoshino** [comment]
i recognise my old AERIALS personalized cache folder path :)

**Comment by glouel** [comment]
So, for some reasons, it looks like your old folder is still taken into account, see this line : 

19:50:49 : cache to be used : Optional("/Volumes/DATAS/AERIALS")

That folder doesn't seem to exist and thus everything fails afterwards. I've added a fix to check if the folder exist before using it, and if not fail back to the folders I specified above. 

That location is persisted in the plist, so, it looks like you didn't erase the plist ? The plist should be in 

/Users/YOURUSERNAME/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist

I would suggest looking for it too in the parent folder (not ByHost), and maybe in /Library/Preferences ? 

In any case, this build should ignore your non existing folder, plist or not : 
https://github.com/glouel/Aerial/releases/download/v1.4.7test4/Aerial.saver.zip

Let me know if that works !

**Comment by hachirotahoshino** [comment]
Thanks, i promess i trashed the byHost plist :) i also didn't understand why it kept the settings, i thought Saved Applications State but found nothing... and you were right, there was also a plist directly below /preferences, should have checked sorry :-/

**Comment by hachirotahoshino** [comment]
i was so sure it should be in ByHost i haven't checked there...

**Comment by glouel** [comment]
No worries, I've seen it a couple of times in the main Preferences instead of ByHost here, but I thought it was something on my development machine.

All good then ?

**Comment by hachirotahoshino** [comment]
Before d/ling your test4 version, i just tried trashing the second plist, no more plist either in prefs or byhost... and same issue. so i am going with the test4 and i'll tell you.

**Comment by hachirotahoshino** [comment]
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2713155/AerialLog.txt)

FYI the log with earlier and latest test3 version now i'll test4

**Comment by hachirotahoshino** [comment]
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2713177/AerialLog.txt)

![capture d ecran 2018-12-27 a 20 31 15](https://user-images.githubusercontent.com/38068857/50492182-6b768b00-0a16-11e9-86ea-4357a0114635.jpg)

![capture d ecran 2018-12-27 a 20 30 54](https://user-images.githubusercontent.com/38068857/50492183-6b768b00-0a16-11e9-8a14-c7ff1ae20863.jpg)

better but not there yet, or is it normal?

**Comment by glouel** [comment]
First, your log looks ok now. 

20:23:49 : cache to be used : Optional("/Users/ssm/Library/Caches/Aerial")

Your first screenshot looks ok, not the second, did you mean to post both ? Was the first launch ok but not the second ? Bit of context would help. Close and restart System Preferences maybe ?

**Comment by hachirotahoshino** [comment]
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2713220/AerialLog.txt)
[aerialtest4.mov.zip](https://github.com/JohnCoates/Aerial/files/2713221/aerialtest4.mov.zip)

sorry i explained badly, video worth thousands words :)

**Comment by glouel** [comment]
No problem!

Just saw your video, that's indeed a new issue. Did you completely close System Preferences (not just go back to the main panel) before reopening it ? I'd suggest using Console as you did above and look for any error thrown there when you click the button.

**Comment by hachirotahoshino** [comment]
yep, done that! quit prefs if i just go back to main panel and come back i go directly to blank page!

i’ll do that tomorrow! gotta go to bed for now! thanks for everything so far, i’ll keep you updated ;)

**Comment by hachirotahoshino** [comment]
`
[C21 Hostname#8d0a2361:443 tcp, url: https://sylvan.apple.com/Aerials/2x/Videos/comp_A105_C003_0212CT_FLARE_v10_SDR_PS_FINAL_20180711_SDR_4K_HEVC.mov, tls] cancelled
	[C21.1 <private> 192.168.69.252:51003<->IPv4#93794966:443]
	Connected Path: satisfied (Path is satisfied), interface: en2, ipv4, dns
	Duration: 0.719s, DNS @0.000s took 0.002s, TCP @0.004s took 0.019s, TLS took 0.063s
	bytes in/out: 3176653/1000, packets in/out: 2196/3, rtt: 0.023s, retransmitted packets: 0, out-of-order packets: 0
`
![capture d ecran 2018-12-28 a 05 37 47](https://user-images.githubusercontent.com/38068857/50502848-c8982e00-0a62-11e9-8fcc-8c6dd0052731.jpg)

**Comment by hachirotahoshino** [comment]
This one loops good news is it started populating the Caches/Aerial folder so it is not right access to the folder:

![capture d ecran 2018-12-28 a 05 41 17](https://user-images.githubusercontent.com/38068857/50502910-39d7e100-0a63-11e9-98de-8958e85ee46a.jpg)

**Comment by hachirotahoshino** [comment]
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2713987/AerialLog.txt)

**Comment by glouel** [comment]
Thanks for you for looking into it, some thoughts: 
- the start/cancel/cancelled on console is normal, it's just the SSL connection being closed after being used. What you saw here was the streaming of a video, which is quite ok. 
- your Cache/Aerial/ folder looks correct indeed (finally !)
- if I read  your log correctly, you opened the "Economiseur d'écran" panel, waited for maybe 30 seconds before clicking the options button, right ? If so, it looks like the panel started correctly (and even started streaming a video, despite not being displayed

Now that I know what happened, I'll try replicating what you did (setting a custom folder then deleting it) to see if I can reproduce the panel not showing up. I'll let you know but we're getting close, and in the meantime it will have fixed a lot of edge cases. 

To answer one of your older questions about not showing an error message, a huge problem with screensavers is that they have to be able to run, no questions asked, no error shown, right after install. When you go into "Economiseur d'écran", the mini preview you see above the button is quite literally macOS launching a mini-scale screensaver. And that's the first thing you see after install (unless you go hot corner and directly launch the screensaver full size, but the result is the same). So we always need to be able to fallback and we don't have the option to prompt the user when something is wrong in screensaver mode (which drives me slightly nuts with the auto update process, as of right now I can only make it prompt you to install the update if you go to the preference panel, and not prompt you, say, after you come back to desktop when waking up your computer).

**Comment by glouel** [comment]
So, found a small issue but I'm not confident that is it. Can I bother you to try this build anyway ? 

https://github.com/glouel/Aerial/releases/download/v1.4.7test5/Aerial.saver.zip

If it still fails, I'd love if you could send me the full console.log, following this procedure : https://support.omnigroup.com/console-osx/

Don't filter by Aerial this time on top left, use the "Effacer" button in the menu bar before you start (will make the log smaller), then go back to click on the button that fails, (give it a few seconds). Then attach me that log. Thanks !

**Comment by hachirotahoshino** [comment]
[console.txt](https://github.com/JohnCoates/Aerial/files/2716201/console.txt)
[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/2716202/AerialLog.txt)

So, i still have blank window when i choose to edit aerial setting so here are the console export and aerial log (with the test5 you sent me)

**Comment by glouel** [comment]
Hey @hachirotahoshino sorry for not getting back to you earlier. 

So I looked at your logs, here's the error : 

```
erreur	09:29:53.098929 +0100	sandboxd	Sandbox: mdwrite(1601) System Policy: deny(1) file-write-xattr /System/Library/PreferencePanes/DesktopScreenEffectsPref.prefPane
```

That's a bit peculiar ! What seems to be happening is that the system tries to change the extended attributes of `DesktopScreenEffectsPref.prefPane`(which I'm pretty sure is the preference panel where you pick the desktop background and which screensaver to use).

Why does the system want to write extended attributes to that pane, I have no idea. And since it's not happening in Aerial's code I can't figure it out outright. Some things I checked here and would suggest you double check : 

First, look if you have any attributes on that folder:
```
glouel@iMac-de-Guillaume-2:~$ ls -l@ /System/Library/PreferencePanes/DesktopScreenEffectsPref.prefPane
total 0
drwxr-xr-x  7 root  wheel  224 20 déc 13:43 Contents 
```
I don't seem to have any, when you have extended attributes, it shows with a @ behind the read/write/execute permissions, here's an example : 

```
glouel@iMac-de-Guillaume-2:~$ ls -l@ thumbnail2.tiff
-rw-r--r--@ 1 glouel  staff  26778 23 oct 17:35 thumbnail2.tiff
	com.apple.lastuseddate#PS	   16
	com.apple.quarantine	   22
```
You can see the @ and the extended attributes.

Some ideas of things you should try :

- you could try 1.4.6 again (the stable build) now that your plist is fixed by running the latest version. Maybe that will work (doubtful but maybe you have that issue because I've added Sparkle to 1.4.7test versions)

- if that doesn't work, check extended attributes on Aerial.saver, I don't have any here : 
```
glouel@iMac-de-Guillaume-2:~$ ls -l@ ~/Library/Screen\ Savers/Aerial.saver
total 0
drwxr-xr-x  7 glouel  staff  224 27 déc 20:04 Contents
```

- Because you mentioned you manually edited the plist, I would suggest checking attributes to that too (again, I don't have any) :
```
glouel@iMac-de-Guillaume-2:~$ ls -l@ ~/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist
-rw-------  1 glouel  staff  60524 28 déc 14:31 /Users/glouel/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist
```

- There are two choices when you install Aerial, all users or only your user, whatever you picked last time, try the other ?

There's probably something that you changed somewhere when you tried to edit things manually that is now tripping the sandbox mechanism, so if you remember what you did, you might want to check that too. You may want at last resort to repair permissions, see here : https://support.apple.com/en-us/HT201560

Let me know what you find.

**Comment by glouel** [comment]
> here the checks:
> `HackThalo:~ ssm$ ls -l@ ~/Library/Screen\ Savers/Aerial.saver`
> `total 0`
> `drwxr-xr-x@ 7 ssm staff 224 28 déc 14:39 Contents`
> `	com.apple.quarantine	57 `
> 
> This looks weird QUARANTINE . it's normal?

I'd say nope, shouldn't be ! 

> gonna try with 1.4.6
> 
> ![capture d ecran 2018-12-31 a 17 26 07](https://user-images.githubusercontent.com/38068857/50563924-35405080-0d21-11e9-8640-443f51fa4a30.jpg)
> 
> Finally! i don't touch anything anymore :)

🎉🎉🎉

My guess, somehow those builds got quarantined and rolling back to 1.4.6 put a fresh non quarantined file as it should be. 

> Thanks a lot
> 
> Bon réveillon et bonne année 2019!

Pas de souci, bon réveillon aussi  et content qu'on ait enfin résolu le truc ;)
